app/scripts/controllers/pipelines.js

```diff
@@ -0,0 +1,140 @@
+    $scope.buildConfigs = {};
+    $scope.statsByConfig = {};
+
+    // get and clear any alerts
+    AlertMessageService.getAlerts().forEach(function(alert) {
+      $scope.alerts[alert.name] = alert.data;
+    });
+    AlertMessageService.clearAlerts();
+
+    var watches = [];
+
+    ProjectsService
+      .get($routeParams.project)
+      .then(_.spread(function(project, context) {
+        $scope.project = project;
+
+        var builds, statsByConfig = {};
+
+        var buildConfigForBuild = $filter('buildConfigForBuild');
+        var isIncomplete = $filter('isIncompleteBuild');
+        var isPipeline = $filter('isJenkinsPipelineStrategy');
+        var isNewer = $filter('isNewerResource');
+
+        var updateStats = function(buildConfigName, build) {
+          if (!statsByConfig[buildConfigName]) {
+            statsByConfig[buildConfigName] = {
+              count: 0,
+              // total duration of all builds for calculating averages
+              totalDuration: 0
+            };
+          }
+          statsByConfig[buildConfigName].count++;
+          statsByConfig[buildConfigName].totalDuration += BuildsService.getDuration(build);
+        };
+
+        var update = function() {
+          // Show running pipelines and the last completed.
+          var interestingBuildsByConfig = {}, latestCompleteByConfig = {};
+
+          _.each(builds, function(build) {
+            // Filter on the client until the server supports fieldSelector on spec.strategy.type.
+            if (!isPipeline(build)) {
+              return;
+            }
+
+            var buildConfigName = buildConfigForBuild(build) || "";
+            // Make sure there is a key for the build config.
+            if (!$scope.buildConfigs[buildConfigName]) {
+              $scope.buildConfigs[buildConfigName] = null;
+            }
+            if (isIncomplete(build)) {
+              _.set(interestingBuildsByConfig, [buildConfigName, build.metadata.name], build);
+            } else if (isNewer(build, latestCompleteByConfig[buildConfigName])) {
+              latestCompleteByConfig[buildConfigName] = build;
+            }
+
+            updateStats(buildConfigName, build);
+          });
+
+          // Add the latest complete build for each config to the interesting builds map.
+          _.each(latestCompleteByConfig, function(build, buildConfigName) {
+            _.set(interestingBuildsByConfig, [buildConfigName, build.metadata.name], build);
+          });
+
+          $scope.interestingBuildsByConfig = interestingBuildsByConfig;
+        };
+
+        // var selectJenkinsPipeline = {
+        //   http: {
+        //     params: {
+        //       fieldSelector: 'spec.strategy.type=JenkinsPipeline'
+        //     }
+        //   }
+        // };
+
+        watches.push(DataService.watch("builds", context, function(buildsData) {
+          $scope.buildsLoaded = true;
+          builds = buildsData.by("metadata.name");
+          update();
```

Comment:
Overview and build config pages might be worse if we're worried about this, and debouncing in those controllers doesn't fully address the problem. The watch callback still results in a digest loop and any ng-repeats in the view. I think we'd need to throttle or debounce in DataService to fix it properly since DataService triggers the digest loop on watch updates. I'm not sure we want to make that change in 1.3, though.

I suggest we profile and see if it's an issue, then look at where we'd need to fix it.